### ISSUE TITLE
Race with deferred DDL changes and GC

### ISSUE THREAD

**Issue by GustavoAngulo** [cmu-db/terrier#526]
The error involves the following schedule (Assume there is a table foo):

```
T0				                T1

BEGIN
write into foo
				                BEGIN
				                drop foo (defers a delete action)
				                COMMIT
COMMIT
T0 removes itself from 
curr_active_txns_
				                GC processes the drop foo deferred action which deallocates foo,
				                because current active txn > T1
ERROR*
```
Error: GC tries to unlink T0, which causes reads to foo’s block layout, which GC already deallocated

Stack frame of error:
```
2019-08-28 17:32:06.343057-0400 recovery_test[47728:5704403] =================================================================
2019-08-28 17:32:06.343510-0400 recovery_test[47728:5704403] ==47728==ERROR: AddressSanitizer: heap-use-after-free on address 0x611000025ec8 at pc 0x000103a972cc bp 0x7ffeec8cac60 sp 0x7ffeec8cac58
2019-08-28 17:32:06.343539-0400 recovery_test[47728:5704403] READ of size 4 at 0x611000025ec8 thread T0
2019-08-28 17:32:06.343567-0400 recovery_test[47728:5704403] pc_0x103a972cb###func_terrier::storage::BlockLayout::NumSlots() const###file_block_layout.h###line_81###obj_(recovery_test:x86_64+0x1007642cb)
2019-08-28 17:32:06.343584-0400 recovery_test[47728:5704403] pc_0x1039adf31###func_terrier::storage::TupleAccessStrategy::AccessWithoutNullCheck(terrier::storage::TupleSlot, terrier::common::StrongTypeAlias<terrier::storage::tags::col_id_t_typedef_tag, unsigned short>) const###file_tuple_access_strategy.h###line_163###obj_(recovery_test:x86_64+0x10067af31)
2019-08-28 17:32:06.343600-0400 recovery_test[47728:5704403] pc_0x1039a4559###func_terrier::storage::DataTable::AtomicallyReadVersionPtr(terrier::storage::TupleSlot, terrier::storage::TupleAccessStrategy const&) const###file_data_table.cpp###line_281###obj_(recovery_test:x86_64+0x100671559)
2019-08-28 17:32:06.343629-0400 recovery_test[47728:5704403] pc_0x1039c1f45###func_terrier::storage::GarbageCollector::TruncateVersionChain(terrier::storage::DataTable*, terrier::storage::TupleSlot, terrier::common::StrongTypeAlias<terrier::transaction::tags::timestamp_t_typedef_tag, unsigned long long>) const###file_garbage_collector.cpp###line_137###obj_(recovery_test:x86_64+0x10068ef45)
2019-08-28 17:32:06.343661-0400 recovery_test[47728:5704403] pc_0x1039bff22###func_terrier::storage::GarbageCollector::ProcessUnlinkQueue()###file_garbage_collector.cpp###line_95###obj_(recovery_test:x86_64+0x10068cf22)
2019-08-28 17:32:06.343686-0400 recovery_test[47728:5704403] pc_0x1039be4f7###func_terrier::storage::GarbageCollector::PerformGarbageCollection()###file_garbage_collector.cpp###line_18###obj_(recovery_test:x86_64+0x10068b4f7)
2019-08-28 17:32:06.343700-0400 recovery_test[47728:5704403] pc_0x10343e33b###func_terrier::storage::GarbageCollectorThread::~GarbageCollectorThread()###file_garbage_collector_thread.h###line_34###obj_(recovery_test:x86_64+0x10010b33b)
2019-08-28 17:32:06.343714-0400 recovery_test[47728:5704403] pc_0x10333fb54###func_terrier::storage::GarbageCollectorThread::~GarbageCollectorThread()###file_garbage_collector_thread.h###line_29###obj_(recovery_test:x86_64+0x10000cb54)
2019-08-28 17:32:06.343862-0400 recovery_test[47728:5704403] pc_0x1033578d6###func_terrier::storage::RecoveryTests_ConcurrentDDLChangesTest_Test::TestBody()###file_recovery_test.cpp###line_756###obj_(recovery_test:x86_64+0x1000248d6)
2019-08-28 17:32:06.343880-0400 recovery_test[47728:5704403] pc_0x103aed9cd###func_void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)###file_gtest.cc###line_2402###obj_(recovery_test:x86_64+0x1007ba9cd)
2019-08-28 17:32:06.343895-0400 recovery_test[47728:5704403] pc_0x103abd87a###func_void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)###file_gtest.cc###line_2438###obj_(recovery_test:x86_64+0x10078a87a)
2019-08-28 17:32:06.343961-0400 recovery_test[47728:5704403] pc_0x103abd7a5###func_testing::Test::Run()###file_gtest.cc###line_2474###obj_(recovery_test:x86_64+0x10078a7a5)
2019-08-28 17:32:06.343986-0400 recovery_test[47728:5704403] pc_0x103abeb9f###func_testing::TestInfo::Run()###file_gtest.cc###line_2656###obj_(recovery_test:x86_64+0x10078bb9f)
2019-08-28 17:32:06.344006-0400 recovery_test[47728:5704403] pc_0x103abfb7e###func_testing::TestCase::Run()###file_gtest.cc###line_2774###obj_(recovery_test:x86_64+0x10078cb7e)
2019-08-28 17:32:06.344024-0400 recovery_test[47728:5704403] pc_0x103ac8357###func_testing::internal::UnitTestImpl::RunAllTests()###file_gtest.cc###line_4649###obj_(recovery_test:x86_64+0x100795357)
2019-08-28 17:32:06.344052-0400 recovery_test[47728:5704403] pc_0x103af1bdd###func_bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)###file_gtest.cc###line_2402###obj_(recovery_test:x86_64+0x1007bebdd)
2019-08-28 17:32:06.344082-0400 recovery_test[47728:5704403] pc_0x103ac7e1a###func_bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)###file_gtest.cc###line_2438###obj_(recovery_test:x86_64+0x100794e1a)
2019-08-28 17:32:06.344109-0400 recovery_test[47728:5704403] pc_0x103ac7cdf###func_testing::UnitTest::Run()###file_gtest.cc###line_4257###obj_(recovery_test:x86_64+0x100794cdf)
2019-08-28 17:32:06.344126-0400 recovery_test[47728:5704403] pc_0x103b11f10###func_RUN_ALL_TESTS()###file_gtest.h###line_2233###obj_(recovery_test:x86_64+0x1007def10)
2019-08-28 17:32:06.344144-0400 recovery_test[47728:5704403] pc_0x103b11ef1###func_main###file_gmock_main.cc###line_53###obj_(recovery_test:x86_64+0x1007deef1)
2019-08-28 17:32:06.344160-0400 recovery_test[47728:5704403] pc_0x7fff7e5413d4###func_start###file_<null>###line_-326330848###obj_(libdyld.dylib:x86_64+0x163d4)
2019-08-28 17:32:06.344176-0400 recovery_test[47728:5704403] 
2019-08-28 17:32:06.344190-0400 recovery_test[47728:5704403] 0x611000025ec8 is located 72 bytes inside of 216-byte region [0x611000025e80,0x611000025f58)
2019-08-28 17:32:06.344218-0400 recovery_test[47728:5704403] freed by thread T4 here:
2019-08-28 17:32:06.344234-0400 recovery_test[47728:5704403] pc_0x104f16192###func_wrap__ZdlPv###file_<null>###line_-326331072###obj_(libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x6e192)
2019-08-28 17:32:06.344252-0400 recovery_test[47728:5704403] pc_0x10383206d###func_terrier::storage::SqlTable::~SqlTable()###file_sql_table.h###line_53###obj_(recovery_test:x86_64+0x1004ff06d)
2019-08-28 17:32:06.344270-0400 recovery_test[47728:5704403] pc_0x103831ff4###func_terrier::storage::SqlTable::~SqlTable()###file_sql_table.h###line_53###obj_(recovery_test:x86_64+0x1004feff4)
2019-08-28 17:32:06.344296-0400 recovery_test[47728:5704403] pc_0x103831fc4###func_terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()::operator()() const###file_database_catalog.cpp###line_641###obj_(recovery_test:x86_64+0x1004fefc4)
2019-08-28 17:32:06.344317-0400 recovery_test[47728:5704403] pc_0x103831eec###func_decltype(std::__1::forward<terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()&>(fp)()) std::__1::__invoke<terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()&>(terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()&&&)###file_type_traits###line_4353###obj_(recovery_test:x86_64+0x1004feeec)
2019-08-28 17:32:06.344350-0400 recovery_test[47728:5704403] pc_0x103831e9c###func_void std::__1::__invoke_void_return_wrapper<void>::__call<terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()&>(terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()&&&)###file___functional_base###line_349###obj_(recovery_test:x86_64+0x1004fee9c)
2019-08-28 17:32:06.344410-0400 recovery_test[47728:5704403] pc_0x103831e6c###func_std::__1::__function::__alloc_func<terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'(), std::__1::allocator<terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()>, void ()>::operator()()###file_functional###line_1527###obj_(recovery_test:x86_64+0x1004fee6c)
2019-08-28 17:32:06.344447-0400 recovery_test[47728:5704403] pc_0x10382e638###func_std::__1::__function::__func<terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'(), std::__1::allocator<terrier::catalog::DatabaseCatalog::DeleteTable(terrier::transaction::TransactionContext*, terrier::common::StrongTypeAlias<terrier::catalog::tags::table_oid_t_typedef_tag, unsigned int>)::$_0::operator()() const::'lambda'()>, void ()>::operator()()###file_functional###line_1651###obj_(recovery_test:x86_64+0x1004fb638)
2019-08-28 17:32:06.344478-0400 recovery_test[47728:5704403] pc_0x1033fd3ad###func_std::__1::__function::__value_func<void ()>::operator()() const###file_functional###line_1799###obj_(recovery_test:x86_64+0x1000ca3ad)
2019-08-28 17:32:06.344498-0400 recovery_test[47728:5704403] pc_0x1033f7364###func_std::__1::function<void ()>::operator()() const###file_functional###line_2347###obj_(recovery_test:x86_64+0x1000c4364)
2019-08-28 17:32:06.344515-0400 recovery_test[47728:5704403] pc_0x1039c09e9###func_terrier::storage::GarbageCollector::ProcessDeferredActions()###file_garbage_collector.cpp###line_129###obj_(recovery_test:x86_64+0x10068d9e9)
2019-08-28 17:32:06.344534-0400 recovery_test[47728:5704403] pc_0x1039be72b###func_terrier::storage::GarbageCollector::PerformGarbageCollection()###file_garbage_collector.cpp###line_27###obj_(recovery_test:x86_64+0x10068b72b)
2019-08-28 17:32:06.344559-0400 recovery_test[47728:5704403] pc_0x10343b145###func_terrier::storage::GarbageCollectorThread::GCThreadLoop()###file_garbage_collector_thread.h###line_70###obj_(recovery_test:x86_64+0x100108145)
2019-08-28 17:32:06.344577-0400 recovery_test[47728:5704403] pc_0x10343af67###func_terrier::storage::GarbageCollectorThread::GarbageCollectorThread(terrier::transaction::TransactionManager*, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >)::'lambda'()::operator()() const###file_garbage_collector_thread.h###line_27###obj_(recovery_test:x86_64+0x100107f67)
2019-08-28 17:32:06.344597-0400 recovery_test[47728:5704403] pc_0x10343aedc###func_decltype(std::__1::forward<terrier::storage::GarbageCollectorThread::GarbageCollectorThread(terrier::transaction::TransactionManager*, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >)::'lambda'()>(fp)()) std::__1::__invoke<terrier::storage::GarbageCollectorThread::GarbageCollectorThread(terrier::transaction::TransactionManager*, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >)::'lambda'()>(terrier::storage::GarbageCollectorThread::GarbageCollectorThread(terrier::transaction::TransactionManager*, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >)::'lambda'()&&)###file_type_traits###line_4353###obj_(recovery_test:x86_64+0x100107edc)
2019-08-28 17:32:06.344642-0400 recovery_test[47728:5704403] pc_0x10343ae84###func_void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, terrier::storage::GarbageCollectorThread::GarbageCollectorThread(terrier::transaction::TransactionManager*, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >)::'lambda'()>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, terrier::storage::GarbageCollectorThread::GarbageCollectorThread(terrier::transaction::TransactionManager*, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >)::'lambda'()>&, std::__1::__tuple_indices<>)###file_thread###line_342###obj_(recovery_test:x86_64+0x100107e84)
2019-08-28 17:32:06.344673-0400 recovery_test[47728:5704403] pc_0x10343a806###func_void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, terrier::storage::GarbageCollectorThread::GarbageCollectorThread(terrier::transaction::TransactionManager*, std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> >)::'lambda'()> >(void*)###file_thread###line_352###obj_(recovery_test:x86_64+0x100107806)
2019-08-28 17:32:06.344714-0400 recovery_test[47728:5704403] pc_0x7fff7e72f2ea###func__pthread_body###file_<null>###line_-326331072###obj_(libsystem_pthread.dylib:x86_64+0x32ea)
2019-08-28 17:32:06.344733-0400 recovery_test[47728:5704403] pc_0x7fff7e732248###func__pthread_start###file_<null>###line_-326331072###obj_(libsystem_pthread.dylib:x86_64+0x6248)
2019-08-28 17:32:06.344750-0400 recovery_test[47728:5704403] pc_0x7fff7e72e40c###func_thread_start###file_<null>###line_-326331072###obj_(libsystem_pthread.dylib:x86_64+0x240c)
2019-08-28 17:32:06.344766-0400 recovery_test[47728:5704403] 
2019-08-28 17:32:06.344800-0400 recovery_test[47728:5704403] previously allocated by thread T0 here:
2019-08-28 17:32:06.344827-0400 recovery_test[47728:5704403] pc_0x104f15b92###func_wrap__Znwm###file_<null>###line_-326331072###obj_(libclang_rt.asan_osx_dynamic.dylib:x86_64h+0x6db92)
2019-08-28 17:32:06.344856-0400 recovery_test[47728:5704403] pc_0x103a46d84###func_terrier::storage::SqlTable::SqlTable(terrier::common::ObjectPool<terrier::storage::RawBlock, terrier::storage::BlockAllocator>*, terrier::catalog::Schema const&)###file_sql_table.cpp###line_53###obj_(recovery_test:x86_64+0x100713d84)
2019-08-28 17:32:06.344876-0400 recovery_test[47728:5704403] pc_0x103a47b54###func_terrier::storage::SqlTable::SqlTable(terrier::common::ObjectPool<terrier::storage::RawBlock, terrier::storage::BlockAllocator>*, terrier::catalog::Schema const&)###file_sql_table.cpp###line_10###obj_(recovery_test:x86_64+0x100714b54)
2019-08-28 17:32:06.344896-0400 recovery_test[47728:5704403] pc_0x103344262###func_terrier::storage::RecoveryTests::CreateTable(terrier::transaction::TransactionContext*, terrier::common::ManagedPointer<terrier::catalog::DatabaseCatalog>, terrier::common::StrongTypeAlias<terrier::catalog::tags::namespace_oid_t_typedef_tag, unsigned int>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)###file_recovery_test.cpp###line_150###obj_(recovery_test:x86_64+0x100011262)
2019-08-28 17:32:06.344932-0400 recovery_test[47728:5704403] pc_0x1033564ba###func_terrier::storage::RecoveryTests_ConcurrentDDLChangesTest_Test::TestBody()###file_recovery_test.cpp###line_730###obj_(recovery_test:x86_64+0x1000234ba)
2019-08-28 17:32:06.345135-0400 recovery_test[47728:5704403] pc_0x103aed9cd###func_void testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)###file_gtest.cc###line_2402###obj_(recovery_test:x86_64+0x1007ba9cd)
2019-08-28 17:32:06.345178-0400 recovery_test[47728:5704403] pc_0x103abd87a###func_void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)###file_gtest.cc###line_2438###obj_(recovery_test:x86_64+0x10078a87a)
2019-08-28 17:32:06.345199-0400 recovery_test[47728:5704403] pc_0x103abd7a5###func_testing::Test::Run()###file_gtest.cc###line_2474###obj_(recovery_test:x86_64+0x10078a7a5)
2019-08-28 17:32:06.345218-0400 recovery_test[47728:5704403] pc_0x103abeb9f###func_testing::TestInfo::Run()###file_gtest.cc###line_2656###obj_(recovery_test:x86_64+0x10078bb9f)
2019-08-28 17:32:06.345235-0400 recovery_test[47728:5704403] pc_0x103abfb7e###func_testing::TestCase::Run()###file_gtest.cc###line_2774###obj_(recovery_test:x86_64+0x10078cb7e)
2019-08-28 17:32:06.345253-0400 recovery_test[47728:5704403] pc_0x103ac8357###func_testing::internal::UnitTestImpl::RunAllTests()###file_gtest.cc###line_4649###obj_(recovery_test:x86_64+0x100795357)
2019-08-28 17:32:06.345271-0400 recovery_test[47728:5704403] pc_0x103af1bdd###func_bool testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)###file_gtest.cc###line_2402###obj_(recovery_test:x86_64+0x1007bebdd)
2019-08-28 17:32:06.345291-0400 recovery_test[47728:5704403] pc_0x103ac7e1a###func_bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)###file_gtest.cc###line_2438###obj_(recovery_test:x86_64+0x100794e1a)
2019-08-28 17:32:06.345309-0400 recovery_test[47728:5704403] pc_0x103ac7cdf###func_testing::UnitTest::Run()###file_gtest.cc###line_4257###obj_(recovery_test:x86_64+0x100794cdf)
2019-08-28 17:32:06.345327-0400 recovery_test[47728:5704403] pc_0x103b11f10###func_RUN_ALL_TESTS()###file_gtest.h###line_2233###obj_(recovery_test:x86_64+0x1007def10)
2019-08-28 17:32:06.345344-0400 recovery_test[47728:5704403] pc_0x103b11ef1###func_main###file_gmock_main.cc###line_53###obj_(recovery_test:x86_64+0x1007deef1)
2019-08-28 17:32:06.345361-0400 recovery_test[47728:5704403] pc_0x7fff7e5413d4###func_start###file_<null>###line_-326331072###obj_(libdyld.dylib:x86_64+0x163d4)
2019-08-28 17:32:06.345377-0400 recovery_test[47728:5704403] 
```

**Comment by GustavoAngulo** [comment]
Adding to this issue, another race exists with a similar schedule because GC will blast away a table before the log serializer is able to copy non-inlined varlens. As @tli2 a possible fix to this is to wait to remove a txn from the `curr_active_txn` set until its logs have been serialized.

**Comment by yash620** [comment]
We identified multiple things we believe would address this. 
1. Much like how we implemented drop index we would have a double deferral on drop table and also defer the unlink calls. This way the table clean up would be executed after the unlink calls. 

2. Another change that seems to have happened is that in a single call to PerformGarbageCollection a timestamps are acquired multiple times, which seems to lead to multiple epochs being processed in a single call. Changing this back to being a single timestamp could address this issue, but @jrolli  suggested that enforcing a stricter semantics now could make it more difficult in the future when we move towards a multi-threaded GC.
